src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -1237,6 +1248,262 @@ def _process_case_value(self, selector, node, node_parent, parent):
+
+    def _where_construct_handler(self, node, parent):
+        '''
+        Construct the canonical PSyIR representation of a WHERE construct or
+        statement.
+
+        :param node: node in the fparser2 parse tree representing the WHERE.
+        :type node: :py:class:`fparser.two.Fortran2003.Where_Construct` or \
+                    :py:class:`fparser.two.Fortran2003.Where_Stmt`
+        :param parent: parent node in the PSyIR.
+        :type parent: :py:class:`psyclone.psyGen.Node`
+
+        :returns: the top-level Loop object in the created loop nest.
+        :rtype: :py:class:`psyclone.psyGen.Loop`
+
+        :raises InternalError: if the parse tree does not have the expected \
+                               structure.
+
+        '''
+        # Check that the fparser2 parse tree has the expected structure
+        if isinstance(node, Fortran2003.Where_Stmt):
+            was_single_stmt = True
+            annotations = ["was_where", "was_single_stmt"]
+            logical_expr = [node.items[0]]
+            if not len(node.items) == 2:
+                raise InternalError(
+                    "Expected a Fortran2003.Where_Stmt to have exactly two "
+                    "entries in 'items' but found {0}: {1}".format(
+                        len(node.items), str(node.items)))
+            if not isinstance(node.items[1], Fortran2003.Assignment_Stmt):
+                raise InternalError(
+                    "Expected the second entry of a Fortran2003.Where_Stmt "
+                    "items tuple to be an Assignment_Stmt but found: {0}".
+                    format(type(node.items[1]).__name__))
+        else:
+            if not isinstance(node.content[0],
+                              Fortran2003.Where_Construct_Stmt):
+                raise InternalError("Failed to find opening where construct "
+                                    "statement in: {0}".format(str(node)))
+            if not isinstance(node.content[-1], Fortran2003.End_Where_Stmt):
+                raise InternalError("Failed to find closing end where "
+                                    "statement in: {0}".format(str(node)))
+            was_single_stmt = False
+            annotations = ["was_where"]
+            logical_expr = node.content[0].items
+        # Examine the logical-array expression (the mask) in order to
+        # determine the number of nested loops required. The Fortran
+        # standard allows bare array notation here (e.g. `a < 0.0` where
+        # `a` is an array) and thus we would need to examine our SymbolTable
+        # to find out the rank of `a`. For the moment we limit support to
+        # the NEMO style where the fact that `a` is an array is made
+        # explicit using the colon notation, e.g. `a(:, :) < 0.0`.
+
+        # Use a basic Schedule as our fake parent as it doesn't have a
+        # SymbolTable (and thus won't check for variable declarations).
```

Comment:
I'm afraid I don't understand this comment - I accept that, as written, the variable declarations will never get checked but that's because we won't currently be able to process (a lot of) NEMO code if we do check for declarations. NEMO makes heavy use of `USE` without an `ONLY` clause and therefore we currently have no way of finding declarations. If you're suggesting a way around this problem then I'm all for it once I understand it!

Reply:
But this is not nemo-specific code and even if it were it is not the right place to be worrying about unresolved externals.

Reply:
Thanks for modifying. Could you please remove the comment starting "Use a basic schedule ..." and replace with an explanation that a temporary parent is used as the code is going to be placed in a different location once the appropriate new loops and if statement have been created.